src/api/accounts.rs

```diff
@@ -0,0 +1,515 @@
+    where
+        S: AsRef<str>,
+        Sig: IntoSignature,
+    {
+        let message_hash = self.hash_message(message);
+        let signature = signature.into_signature();
+
+        let public_key = signature.recover(&message_hash[..]).map_err(EthsignError::from)?;
+
+        Ok(public_key.address().into())
+    }
+}
+
+type MaybeReady<T, R> = Either<FutureResult<R, Error>, CallFuture<R, <T as Transport>::Out>>;
+
+type TxParams<T> = Join3<MaybeReady<T, U256>, MaybeReady<T, U256>, MaybeReady<T, String>>;
+
+/// Future resolving when transaction signing is complete
+pub struct SignTransactionFuture<T: Transport> {
+    tx: TransactionParameters,
+    key: SecretKey,
+    inner: TxParams<T>,
+}
+
+impl<T: Transport> SignTransactionFuture<T> {
+    /// Creates a new SignTransactionFuture with accounts and transaction data.
+    pub fn new(accounts: &Accounts<T>, tx: TransactionParameters, key: &SecretKey) -> SignTransactionFuture<T> {
+        macro_rules! maybe {
+            ($o: expr, $f: expr) => {
+                match $o.clone() {
+                    Some(value) => Either::A(future::ok(value)),
+                    None => Either::B($f),
+                }
+            };
+        }
+
+        let from = key.public().address().into();
+        let inner = Future::join3(
+            maybe!(tx.nonce, accounts.web3().eth().transaction_count(from, None)),
+            maybe!(tx.gas_price, accounts.web3().eth().gas_price()),
+            maybe!(None, accounts.web3().net().version()),
+        );
+
+        SignTransactionFuture {
+            tx,
+            key: key.clone(),
+            inner,
+        }
+    }
+}
+
+impl<T: Transport> Future for SignTransactionFuture<T> {
+    type Item = SignedTransaction;
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let (nonce, gas_price, chain_id) = try_ready!(self.inner.poll());
+
+        let chain_id = chain_id.parse::<u64>().map_err(|e| Error::Decoder(e.to_string()))?;
+        let tx = RawTxParameters {
+            nonce,
+            to: self.tx.to.unwrap_or_default(),
+            value: self.tx.value.unwrap_or_default(),
+            gas_price,
+            gas: self.tx.gas.unwrap_or(TRANSACTION_BASE_GAS_FEE),
+            data: self.tx.data.take().unwrap_or_default(),
+        };
+
+        let signed = tx.sign(&self.key, Some(chain_id))?;
+
+        Ok(Async::Ready(signed))
+    }
+}
+
+/// Trait for converting data into an `ethsign::Signature`.
+pub trait IntoSignature {
+    fn into_signature(self) -> Signature;
+}
+
+impl IntoSignature for (u8, H256, H256) {
+    fn into_signature(self) -> Signature {
+        let (v, r, s) = self;
+        Signature {
+            v: v - 27, // ¯\_(ツ)_/¯
+            r: r.into(),
+            s: s.into(),
+        }
+    }
+}
+
+impl<'a> IntoSignature for &'a Bytes {
+    fn into_signature(self) -> Signature {
+        if self.0.len() != 65 {
+            panic!("invalid signature bytes");
```

Comment:
Alright, that sounds like a good idea.

Reply:
@tomusdrw Actually, thinking about it now it would be nice to be able to do something like:
```rust
let signed_data = accounts.sign("hello world", &key);
let address = accounts.recover(signed_data)?;
assert_eq!(address, key.public().address());
let signed_tx = accounts.sign(tx, &key).await?;
let address = accounts.recover(signed_tx)?;
assert_eq!(address, key.public().address());
```

What do you think?